mart/callbacks/anomalib.py

```diff
@@ -0,0 +1,425 @@
+
+        pbar.close()
+
+        # Save best batch and history
+        for key, value in best_batch.items():
+            self.best[key].append(value)
+        for key, value in batch_history.items():
+            if isinstance(value[0], torch.Tensor):
+                value = torch.stack(value, dim=-1)
+            self.history[key].append(value)
+
+        # Update batch with items from best batch
+        for key in batch:
+            value = best_batch[key]
+            if isinstance(value, torch.Tensor):
+                value = value.to(device)
+            batch[key] = value
+
+    def on_test_epoch_end(self, trainer, pl_module):
+        results = {
+            "best": flatten(self.best),
+            "history": flatten(self.history),
+        }
+
+        torch.save(results, os.path.join(trainer.default_root_dir, "results.pt"))
+        del self.best, self.history
+
+
+def flatten(dict_of_lists):
+    # Flatten list of tensors/lists into tensor/list
+    flattened = {}
+
+    for key, value in dict_of_lists.items():
+        if isinstance(value[0], torch.Tensor) and value[0].ndim > 0:
+            value = torch.concat(value)
+        elif isinstance(value[0], torch.Tensor):
+            value = torch.stack(value)
+        elif isinstance(value[0], list):
+            value = sum(value, [])
+        flattened[key] = value
+
+    return flattened
+
+
+# FIXME: Make this a @staticmethod
+def perturb_image(
+    image,
+    *,
+    angle=None,
+    hue=None,
+    sat=None,
+    image_mean=None,
+    image_std=None,
+    mode="three_pass",  # bilinear | nearest | three_pass
+    padding_mode="replicate",  # constant | replicate | reflect | circular | zeros | border | reflection
+    **kwargs,
+):
+    image_adv = image.clone()
+
+    # Return image to [0, 1] space
+    if image_mean is not None and image_std is not None:
+        image_adv = image_adv * image_std[..., None, None] + image_mean[..., None, None]
+
+    # Rotate image by angle degrees
+    if angle is not None:
+        if mode == "three_pass":
+            theta = torch.deg2rad(angle)
+            image_adv = rotate_three_pass(
+                image_adv, theta, N=-1, padding_mode=padding_mode
+            )
+        else:
+            image_adv = rotate(image_adv, angle, mode=mode, padding_mode=padding_mode)
+
+    # Convert image from RGB to HSV space
+    if hue is not None or sat is not None:
+        image_hsv = rgb_to_hsv(image_adv)
+        image_hue, image_sat, image_val = torch.unbind(image_hsv, dim=-3)
+
+    # Additive hue perturbation with STE clipping
+    if hue is not None:
+        image_hue = image_hue + hue[:, None, None]
+        image_hue = torch.remainder(image_hue, 2 * torch.pi)
+
+    # Additive saturation perturbation with STE clipping
+    if sat is not None:
+        image_sat = image_sat + sat[:, None, None]
+        image_sat = image_sat + (torch.clip(image_sat, 0.0, 1.0) - image_sat).detach()
+
+    # Convert image fro HSV to RGB space
```

Comment:
"from"

Reply:
Fixed in faefb8e53dfb9d7813f81436ccfd41e589f31317.